File: src/nodding_lidar/nodding_lidar/scan_to_3d.py
#!/usr/bin/env python3
"""
scan_to_3d.py - Converts 2D RPLidar scans + nodding angle into 3D pointcloud
Core for NIDAR AirMouse: 2D->3D via oscillation

Input:
- /scan (LaserScan) from RPLidar, 360° horizontal
- /servo_angle / /joint_states (pitch angle of lidar mount)
- /offboard/pose or /camera/odom (drone pose for world frame)

Output:
- /scan_3d (PointCloud2) - accumulated 3D points in lidar frame + world frame
- /map_3d (OctoMap style) for FUEL planner
- /scan (still publish 2D projection for 2D mapping)

Algorithm (nodding lidar):
For each 2D scan point at (r, theta) in lidar's local horizontal plane,
with mount pitch = phi (servo angle):
- Point in mount frame: x = r*cos(theta), y = r*sin(theta), z=0
- Rotate by pitch phi around Y axis (nodding): 
  x' = x*cos(phi) + z*sin(phi)
  y' = y
  z' = -x*sin(phi) + z*cos(phi)
- Then transform by drone pose to world frame

Accumulate over full nodding sweep (e.g., -45° to +45°) to get dense 3D scan of room.
For 15x15m maze, 8ft height, this captures floor, walls, ceiling, survivors.

Also generates 2D projection for 2D map (required for NIDAR scoring).
"""
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan, PointCloud2, PointField, JointState
from geometry_msgs.msg import PoseStamped
from std_msgs.msg import Float64
import numpy as np
import math
import struct
from collections import deque

class ScanTo3D(Node):
    def __init__(self):
        super().__init__('scan_to_3d')
        self.declare_parameter('scan_topic', '/scan')
        self.declare_parameter('servo_angle_topic', '/servo_angle_cmd')
        self.declare_parameter('accumulate_scans', 20)  # number of 2D scans per 3D cloud
        self.declare_parameter('min_range', 0.15)
        self.declare_parameter('max_range', 12.0)
        self.declare_parameter('publish_2d_projection', True)

        self.scan_topic = self.get_parameter('scan_topic').value
        self.servo_topic = self.get_parameter('servo_angle_topic').value
        self.accumulate_n = self.get_parameter('accumulate_scans').value
        self.min_range = self.get_parameter('min_range').value
        self.max_range = self.get_parameter('max_range').value

        self.current_pitch = 0.0
        self.current_pose = None
        self.scan_buffer = deque(maxlen=self.accumulate_n)

        # Subs
        self.scan_sub = self.create_subscription(LaserScan, self.scan_topic, self.scan_callback, 10)
        self.pitch_sub = self.create_subscription(Float64, self.servo_topic, self.pitch_callback, 10)
        self.joint_sub = self.create_subscription(JointState, '/joint_states', self.joint_callback, 10)
        self.pose_sub = self.create_subscription(PoseStamped, '/offboard/pose', self.pose_callback, 10)

        # Pubs
        self.cloud_pub = self.create_publisher(PointCloud2, '/scan_3d', 10)
        self.cloud_world_pub = self.create_publisher(PointCloud2, '/scan_3d_world', 10)
        self.scan_2d_pub = self.create_publisher(LaserScan, '/scan_2d_projected', 10)

        self.get_logger().info(f"ScanTo3D: {self.accumulate_n} scans per 3D cloud, pitch from {self.servo_topic}")

    def pitch_callback(self, msg):
        self.current_pitch = msg.data

    def joint_callback(self, msg):
        try:
            idx = msg.name.index('rplidar_pitch_joint')
            self.current_pitch = msg.position[idx]
        except ValueError:
            pass

    def pose_callback(self, msg):
        self.current_pose = msg

    def scan_callback(self, msg):
        # Store scan with current pitch
        self.scan_buffer.append({
            'scan': msg,
            'pitch': self.current_pitch,
            'pose': self.current_pose
        })

        # When buffer full, generate 3D cloud
        if len(self.scan_buffer) >= self.accumulate_n:
            self.generate_3d_cloud()
            # Optionally clear buffer for next sweep
            # self.scan_buffer.clear()  # Keep sliding window

    def generate_3d_cloud(self):
        points_local = []  # in base_link frame
        points_world = []

        for entry in self.scan_buffer:
            scan = entry['scan']
            pitch = entry['pitch']
            pose = entry['pose']

            # Precompute pitch rotation
            cos_p = math.cos(pitch)
            sin_p = math.sin(pitch)

            angle = scan.angle_min
            for r in scan.ranges:
                if math.isinf(r) or math.isnan(r) or r < self.min_range or r > self.max_range:
                    angle += scan.angle_increment
                    continue

                # 2D point in lidar horizontal plane (lidar frame)
                # RPLidar: 0 angle = forward, CCW
                x_lidar = r * math.cos(angle)
                y_lidar = r * math.sin(angle)
                z_lidar = 0.0

                # Rotate by pitch (nodding) around Y axis
                # Mount pitch rotates lidar forward up/down
                # x' = x*cos(pitch) + z*sin(pitch)
                # z' = -x*sin(pitch) + z*cos(pitch)
                x_mount = x_lidar * cos_p + z_lidar * sin_p
                y_mount = y_lidar
                z_mount = -x_lidar * sin_p + z_lidar * cos_p

                # Transform to base_link: assume lidar mounted at (0,0,0.12) + pitch joint
                # For simplicity, add mount offset
                x_base = x_mount
                y_base = y_mount
                z_base = z_mount + 0.12  # lidar height

                points_local.append([x_base, y_base, z_base])

                # Transform to world if pose available
                if pose is not None:
                    # Simple ENU transform: rotate by yaw and translate
                    q = pose.pose.orientation
                    siny_cosp = 2*(q.w*q.z + q.x*q.y)
                    cosy_cosp = 1 - 2*(q.y*q.y + q.z*q.z)
                    yaw = math.atan2(siny_cosp, cosy_cosp)

                    cos_y = math.cos(yaw)
                    sin_y = math.sin(yaw)

                    x_world = pose.pose.position.x + x_base * cos_y - y_base * sin_y
                    y_world = pose.pose.position.y + x_base * sin_y + y_base * cos_y
                    z_world = pose.pose.position.z + z_base

                    points_world.append([x_world, y_world, z_world])

                angle += scan.angle_increment

        # Publish local cloud
        if len(points_local) > 0:
            self.publish_cloud(points_local, self.cloud_pub, frame_id="base_link")

        # Publish world cloud (for mapping)
        if len(points_world) > 0:
            self.publish_cloud(points_world, self.cloud_world_pub, frame_id="map")

        self.get_logger().info(f"Generated 3D cloud: {len(points_local)} points from {len(self.scan_buffer)} scans, pitch range {min([e['pitch'] for e in self.scan_buffer]):.2f} to {max([e['pitch'] for e in self.scan_buffer]):.2f} rad")

    def publish_cloud(self, points, publisher, frame_id="map"):
        # Downsample if too many
        if len(points) > 10000:
            points = points[::len(points)//10000]

        msg = PointCloud2()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = frame_id
        msg.height = 1
        msg.width = len(points)
        msg.fields = [
            PointField(name='x', offset=0, datatype=PointField.FLOAT32, count=1),
            PointField(name='y', offset=4, datatype=PointField.FLOAT32, count=1),
            PointField(name='z', offset=8, datatype=PointField.FLOAT32, count=1),
        ]
        msg.is_bigendian = False
        msg.point_step = 12
        msg.row_step = msg.point_step * len(points)
        msg.is_dense = True
        buffer = []
        for p in points:
            buffer.append(struct.pack('fff', float(p[0]), float(p[1]), float(p[2])))
        msg.data = b''.join(buffer)
        publisher.publish(msg)

def main(args=None):
    rclpy.init(args=args)
    node = ScanTo3D()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/nodding_lidar_cpp/src/scan_to_3d.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/laser_scan.hpp>
#include <sensor_msgs/msg/point_cloud2.hpp>
#include <sensor_msgs/msg/joint_state.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <std_msgs/msg/float64.hpp>
#include <vector>
#include <deque>
#include <cmath>
#include <cstring>

struct ScanEntry {
  sensor_msgs::msg::LaserScan scan;
  double pitch;
  geometry_msgs::msg::PoseStamped pose;
  bool has_pose;
};

class ScanTo3D : public rclcpp::Node {
public:
  ScanTo3D() : Node("scan_to_3d") {
    this->declare_parameter("accumulate_scans", 20);
    this->declare_parameter("min_range", 0.15);
    this->declare_parameter("max_range", 12.0);

    accumulate_n_ = this->get_parameter("accumulate_scans").as_int();
    min_range_ = this->get_parameter("min_range").as_double();
    max_range_ = this->get_parameter("max_range").as_double();

    scan_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>("/scan", 10, std::bind(&ScanTo3D::scanCb, this, std::placeholders::_1));
    pitch_sub_ = this->create_subscription<std_msgs::msg::Float64>("/servo_angle_cmd", 10, std::bind(&ScanTo3D::pitchCb, this, std::placeholders::_1));
    joint_sub_ = this->create_subscription<sensor_msgs::msg::JointState>("/joint_states", 10, std::bind(&ScanTo3D::jointCb, this, std::placeholders::_1));
    pose_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>("/offboard/pose", 10, std::bind(&ScanTo3D::poseCb, this, std::placeholders::_1));

    cloud_pub_ = this->create_publisher<sensor_msgs::msg::PointCloud2>("/scan_3d", 10);
    cloud_world_pub_ = this->create_publisher<sensor_msgs::msg::PointCloud2>("/scan_3d_world", 10);

    RCLCPP_INFO(this->get_logger(), "C++ ScanTo3D accumulate %d", accumulate_n_);
  }

private:
  void pitchCb(const std_msgs::msg::Float64::SharedPtr msg) { current_pitch_=msg->data; }
  void jointCb(const sensor_msgs::msg::JointState::SharedPtr msg) {
    for (size_t i=0;i<msg->name.size();++i) if (msg->name[i]=="rplidar_pitch_joint") current_pitch_=msg->position[i];
  }
  void poseCb(const geometry_msgs::msg::PoseStamped::SharedPtr msg) { current_pose_=*msg; has_pose_=true; }

  void scanCb(const sensor_msgs::msg::LaserScan::SharedPtr msg) {
    ScanEntry e; e.scan=*msg; e.pitch=current_pitch_; e.pose=current_pose_; e.has_pose=has_pose_;
    buffer_.push_back(e);
    if ((int)buffer_.size() > accumulate_n_) buffer_.pop_front();
    if ((int)buffer_.size() >= accumulate_n_) generate();
  }

  void generate() {
    std::vector<std::array<float,3>> local_pts, world_pts;
    local_pts.reserve(10000); world_pts.reserve(10000);

    for (auto &entry: buffer_) {
      double pitch=entry.pitch;
      double cos_p=std::cos(pitch), sin_p=std::sin(pitch);
      double angle=entry.scan.angle_min;
      double yaw=0;
      if (entry.has_pose) {
        auto q=entry.pose.pose.orientation;
        double siny=2*(q.w*q.z+q.x*q.y);
        double cosy=1-2*(q.y*q.y+q.z*q.z);
        yaw=std::atan2(siny,cosy);
      }
      double cos_y=std::cos(yaw), sin_y=std::sin(yaw);

      for (size_t i=0;i<entry.scan.ranges.size();++i) {
        float r=entry.scan.ranges[i];
        if (std::isinf(r)||std::isnan(r)||r<min_range_||r>max_range_) { angle+=entry.scan.angle_increment; continue; }
        double x_lidar = r*std::cos(angle);
        double y_lidar = r*std::sin(angle);
        double x_mount = x_lidar*cos_p;
        double y_mount = y_lidar;
        double z_mount = -x_lidar*sin_p;

        double x_base=x_mount, y_base=y_mount, z_base=z_mount+0.12;
        local_pts.push_back({(float)x_base,(float)y_base,(float)z_base});

        if (entry.has_pose) {
          double x_world = entry.pose.pose.position.x + x_base*cos_y - y_base*sin_y;
          double y_world = entry.pose.pose.position.y + x_base*sin_y + y_base*cos_y;
          double z_world = entry.pose.pose.position.z + z_base;
          world_pts.push_back({(float)x_world,(float)y_world,(float)z_world});
        }
        angle+=entry.scan.angle_increment;
      }
    }

    publishCloud(local_pts, cloud_pub_, "base_link");
    publishCloud(world_pts, cloud_world_pub_, "map");

    if (!buffer_.empty()) {
      double min_p=buffer_[0].pitch, max_p=buffer_[0].pitch;
      for (auto &e: buffer_) { min_p=std::min(min_p,e.pitch); max_p=std::max(max_p,e.pitch); }
      RCLCPP_INFO(this->get_logger(), "3D cloud %zu local %zu world pitch %.2f to %.2f rad", local_pts.size(), world_pts.size(), min_p, max_p);
    }
  }

  void publishCloud(const std::vector<std::array<float,3>>& pts, rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr pub, const std::string& frame) {
    if (pts.empty()) return;
    // Downsample to 10k
    std::vector<std::array<float,3>> filtered;
    if (pts.size()>10000) {
      int step=pts.size()/10000;
      for (size_t i=0;i<pts.size();i+=step) filtered.push_back(pts[i]);
    } else filtered=pts;

    sensor_msgs::msg::PointCloud2 cloud;
    cloud.header.stamp=this->now(); cloud.header.frame_id=frame;
    cloud.height=1; cloud.width=filtered.size();
    cloud.fields.resize(3);
    cloud.fields[0].name="x"; cloud.fields[0].offset=0; cloud.fields[0].datatype=7; cloud.fields[0].count=1;
    cloud.fields[1].name="y"; cloud.fields[1].offset=4; cloud.fields[1].datatype=7; cloud.fields[1].count=1;
    cloud.fields[2].name="z"; cloud.fields[2].offset=8; cloud.fields[2].datatype=7; cloud.fields[2].count=1;
    cloud.point_step=12; cloud.row_step=cloud.point_step*filtered.size(); cloud.is_dense=true;
    cloud.data.resize(cloud.row_step);
    for (size_t i=0;i<filtered.size();++i) {
      std::memcpy(&cloud.data[i*12], &filtered[i][0], 4);
      std::memcpy(&cloud.data[i*12+4], &filtered[i][1], 4);
      std::memcpy(&cloud.data[i*12+8], &filtered[i][2], 4);
    }
    pub->publish(cloud);
  }

  int accumulate_n_;
  double min_range_, max_range_, current_pitch_=0;
  geometry_msgs::msg::PoseStamped current_pose_;
  bool has_pose_=false;
  std::deque<ScanEntry> buffer_;

  rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr scan_sub_;
  rclcpp::Subscription<std_msgs::msg::Float64>::SharedPtr pitch_sub_;
  rclcpp::Subscription<sensor_msgs::msg::JointState>::SharedPtr joint_sub_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr pose_sub_;
  rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr cloud_pub_, cloud_world_pub_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ScanTo3D>());
  rclcpp::shutdown();
  return 0;
}
